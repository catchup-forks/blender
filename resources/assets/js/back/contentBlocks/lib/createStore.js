import axios from 'axios';
import Vue from 'vue';

const Store = {

    data() {
        return {
            collection: '',
            createUrl: '',
            model: '',
            data: {},
            blocks: [],
        };
    },

    computed: {
        export() {
            return JSON.stringify(
                this.blocks.filter(b => b.markedForRemoval === false)
            );
        },
    },

    methods: {
        init({ collection, createUrl, model, data, initial }) {
            this.collection = collection;
            this.createUrl = createUrl;
            this.model = model;
            this.data = data;

            this.addBlocks(initial);
        },

        addBlocks(blocks) {
            this.blocks = [
                ...this.blocks,
                ...blocks.map(b => ({ ...b, markedForRemoval: false })),
            ];
        },

        async createBlock() {
            const { data: block } = await axios.post(this.createUrl, {
                model_name: this.model.name,
                model_id: this.model.id,
                collection_name: this.collection,
            });
            
            this.addBlocks([block]);
        },
    },
};

function createStore() {
    return new Vue(Store);
}

export default createStore;