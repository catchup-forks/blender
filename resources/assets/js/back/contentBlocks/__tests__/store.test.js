import createStore from '../lib/createStore';

describe('Store', () => {

    let store;

    beforeEach(() => {
        store = createStore();
        store.init({
            collection: 'default',
            createUrl: 'http://blender.dev/blender/api/contentblocks',
            mediaUrl: 'http://blender.dev/blender/api/media',
            model: {
                name: '\\App\\Models\\Article',
                id: 1,
            },
            data: {
                locales: ['nl', 'en'],
                contentLocale: 'nl',
            },
            initial: [],
        });
    });

    it('can be initialized', () => {
        expect(store.$data).toMatchSnapshot();
    });

    it('can add an array of existing blocks', () => {

        const block = {
            id: 1,
            title: 'Lego',
            layout: 'image-left',
        };

        store.addBlocks([block]);

        expect(store.blocks).toMatchSnapshot();
    });

    it('can create a new block', async () => {
        
        await store.createBlock();

        expect(store.blocks).toMatchSnapshot();
    });
});