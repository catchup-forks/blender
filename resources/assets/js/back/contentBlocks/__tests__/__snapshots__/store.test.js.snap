exports[`Store can add an array of existing blocks 1`] = `
Array [
  Object {
    "id": 1,
    "layout": "image-left",
    "markedForRemoval": false,
    "title": "Lego",
  },
]
`;

exports[`Store can be initialized 1`] = `
Object {
  "blocks": Array [],
  "collection": "default",
  "createUrl": "http://blender.dev/blender/api/contentblocks",
  "data": Object {
    "contentLocale": "nl",
    "locales": Array [
      "nl",
      "en",
    ],
  },
  "model": Object {
    "id": 1,
    "name": "\\App\\Models\\Article",
  },
}
`;

exports[`Store can create a new block 1`] = `
Array [
  Object {
    "markedForRemoval": false,
  },
]
`;
